macros.h

```diff
@@ -1,319 +1,438 @@
-  if (argc > cb_pos) { \
-    task->type = sn_async_task_callback; \
-    promise = NULL; \
-    napi_value cb = argv[cb_pos]; \
-    napi_valuetype type; \
-    SN_STATUS_THROWS(napi_typeof(env, cb, &type), "") \
-    if (type != napi_function) { \
-      napi_throw_error(env, "EINVAL", "Callback must be a function"); \
-      return NULL; \
-    } \
-    SN_STATUS_THROWS(napi_create_reference(env, cb, 1, &task->cb), "") \
-  } else { \
-    task->type = sn_async_task_promise; \
-    napi_create_promise(env, &task->deferred, &promise); \
-  }
+    task->req = (void *) req; \
+    js_value_t *promise; \
+    if (argc > cb_pos) { \
+      task->type = sn_async_task_callback; \
+      promise = NULL; \
+      err = js_create_reference(env, argv[cb_pos], 1, &task->cb); \
+      assert(err == 0); \
+    } else { \
+      task->type = sn_async_task_promise; \
+      err = js_create_promise(env, &task->deferred, &promise); \
+      assert(err == 0); \
+    }
 
 #define SN_ASYNC_COMPLETE(message) \
-  napi_value argv[1]; \
+  js_value_t *argv[1]; \
   switch (task->type) { \
   case sn_async_task_promise: { \
     if (task->code == 0) { \
-      napi_get_null(req->env, &argv[0]); \
-      napi_resolve_deferred(req->env, task->deferred, argv[0]); \
+      err = js_get_null(req->env, &argv[0]); \
+      assert(err == 0); \
+      err = js_resolve_deferred(req->env, task->deferred, argv[0]); \
+      assert(err == 0); \
     } else { \
-      napi_value err_msg; \
-      napi_create_string_utf8(req->env, #message, NAPI_AUTO_LENGTH, &err_msg); \
-      napi_create_error(req->env, NULL, err_msg, &argv[0]); \
-      napi_reject_deferred(req->env, task->deferred, argv[0]); \
+      js_value_t *err_msg; \
+      assert(err == 0); \
+      err = js_create_string_utf8(req->env, (const utf8_t *) #message, SN_LIT_LENGTH(message), &err_msg); \
+      assert(err == 0); \
+      err = js_create_error(req->env, NULL, err_msg, &argv[0]); \
+      assert(err == 0); \
+      err = js_reject_deferred(req->env, task->deferred, argv[0]); \
+      assert(err == 0); \
     } \
     task->deferred = NULL; \
     break; \
   } \
   case sn_async_task_callback: { \
     SN_CALLBACK_CHECK_FOR_ERROR(#message) \
-    napi_value callback; \
-    napi_get_reference_value(req->env, task->cb, &callback); \
-    napi_value return_val; \
+    js_value_t *callback; \
+    err = js_get_reference_value(req->env, task->cb, &callback); \
+    assert(err == 0); \
+    js_value_t *return_val; \
     SN_CALL_FUNCTION(req->env, global, callback, 1, argv, &return_val) \
-    napi_close_handle_scope(req->env, scope); \
-    napi_delete_reference(req->env, task->cb); \
+    err = js_delete_reference(req->env, task->cb); \
+    assert(err == 0); \
     break; \
   } \
   }
+
+#define SN_TYPEDARRAY_VIEW(var) \
+  void *var##_data; \
+  size_t var##_size; \
+  js_typedarray_view_t *var##_view; \
+  err = js_get_typedarray_view(env, var, NULL, &var##_data, &var##_size, &var##_view); \
```

Comment:
is it deliberate that we don't get the type info here? this seems like it might break for types larger than 8bit

Reply:
no; i had just misunderstood the intention, good catch non uint8 typedarray support added now

Reply:
Nice! Per the discussion between @kasperisager and @mafintosh i think we can actually drop all width support. Can you update all the typedarray macros (including the existing ones) to remove  `name##_width` variables and  maybe add `assert(name##_type == js_uint8array`?

Reply:
and then can also remove `name##_length` and pass `name##_size` to js_get_typedarray_info/view directly

Reply:
we decided to fix in 5.0.1 to move this fwd